drivers/ma_yya_report/app/models/ma_yya_report/universe_calculator.rb

```diff
@@ -196,19 +189,56 @@ def calculate(&post_processor)
         map(&:CurrentLivingSituation)
     end
 
+    # Most recent Zip for each source client
     private def zip_codes(client)
-      client.source_hmis_clients.map(&:processed_youth_current_zip).compact.uniq || []
+      hmis_source_clients_for(client).map do |hmis_client|
+        hmis_client.addresses.sort(:DateUpdated).pluck(:postal_code).compact.last
+      end.uniq
     end
 
+    # All flex funds received by all source clients (e.g. ["Transportation", "Child care", "Move-in"])
+    # NOTE: Assumes that there is only 1 HMIS data source
     private def flex_funds(client)
-      client.source_hmis_forms.
-        within_range(@filter.start_date .. @filter.end_date).
-        where(name: 'Flex Funds').
-        map(&:flex_funds).flatten.uniq
+      return [] unless flex_funds_service_type && flex_funds_types_cded
+
+      hmis_personal_ids = hmis_source_clients_for(client).pluck(:personal_id)
+      # IDs for all Flex Fund services rendered to the client(s)
+      flex_funds_service_ids = Hmis::Hud::CustomService.hmis.where(personal_id: hmis_personal_ids, service_type: flex_funds_service_type).pluck(:id)
+      # CustomDataElements tied to these CustomServices indicate which fund types were provided
+      flex_funds_types_cded.values.
+        where(owner_id: flex_funds_service_ids).
+        pluck(:value_string).uniq.
+        # Remove parenthesized text from "Move-in (Including first, last and security)"
+        map { |ff_type| ff_type.gsub(/\([^()]*\)/, '').strip }.uniq
     end
 
+    # Valid options are [nil, "English", "Spanish", "Other"]
     private def language(client)
-      client.source_hmis_clients.map(&:primary_language).detect(&:present?)
+      client.source_clients.map do |source_client|
+        # Choose the most recent enrollment that has language information. Sort by entry date since language is collected at entry.
+        enrollment = source_client.enrollments.where.not(translation_needed: nil).order(:entry_date).last
```

Comment:
Should this be limited to enrollments visible to the user?  Feels slightly odd to pull from all of a client's enrollments, but given the data returned, maybe that's ok?

Reply:
I don't know.. it doesn't look like the Flex Funds were ever limited by enrollment visibility, so I don't think this should be either, but I can make a note to clarify next time.

Reply:
It probably doesn't matter, given the scope of the report, but if you could drop a note that it's not limited, that seems like a good middle ground.  Thanks!